hazelcast/src/test/java/com/hazelcast/journal/AbstractEventJournalExpiringTest.java

```diff
@@ -0,0 +1,171 @@
+/*
+ * Copyright (c) 2008-2018, Hazelcast, Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.journal;
+
+import com.hazelcast.config.Config;
+import com.hazelcast.config.EventJournalConfig;
+import com.hazelcast.core.ExecutionCallback;
+import com.hazelcast.core.HazelcastInstance;
+import com.hazelcast.core.ICompletableFuture;
+import com.hazelcast.projection.Projection;
+import com.hazelcast.projection.Projections;
+import com.hazelcast.ringbuffer.ReadResultSet;
+import com.hazelcast.spi.properties.GroupProperty;
+import com.hazelcast.test.HazelcastTestSupport;
+import com.hazelcast.util.function.Predicate;
+import org.junit.Test;
+
+import java.util.Random;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.concurrent.locks.LockSupport;
+
+/**
+ * Base class for implementing data-structure specific event journal test where the journal is expiring.
+ *
+ * @param <EJ_TYPE> the type of the event journal event
+ */
+public abstract class AbstractEventJournalExpiringTest<EJ_TYPE> extends HazelcastTestSupport {
+    private static final Random RANDOM = new Random();
+
+    protected HazelcastInstance[] instances;
+
+    private int partitionId;
+    private TruePredicate<EJ_TYPE> TRUE_PREDICATE = new TruePredicate<EJ_TYPE>();
+    private Projection<EJ_TYPE, EJ_TYPE> IDENTITY_PROJECTION = Projections.identity();
+
+    private void init() {
+        instances = createInstances();
+        partitionId = 1;
+        warmUpPartitions(instances);
+    }
+
+    @Override
+    protected Config getConfig() {
+        int defaultPartitionCount = Integer.parseInt(GroupProperty.PARTITION_COUNT.getDefaultValue());
+        EventJournalConfig eventJournalConfig = new EventJournalConfig()
+                .setEnabled(true)
+                .setMapName("default")
+                .setCacheName("default")
+                .setTimeToLiveSeconds(1)
+                .setCapacity(500 * defaultPartitionCount);
+
+        return smallInstanceConfig()
+                .addEventJournalConfig(eventJournalConfig);
+    }
+
+    @Test
+    public void skipsEventsWhenExpired() throws Throwable {
```

Comment:
The test doesn't fail when I reverted `EventJournalReadOperation` to the original version. Does it test the issue? I've run it twice.

It also takes quite long, almost 3 minutes.

Reply:
The scenario depends on timing so you need to run it many times to hit the issue. And the duration of the test increases the chances we will hit the issue. That's why the test is marked as `SlowTest`.